### DISCUSSION TITLE
How does libvips read the TIFF data of each band？

### DISCUSSION THREAD

**Discussion by Zhaoyazhou1996** [libvips/libvips#2272] [Q&A] [answered]
Hi, I recently tested the speed of GDAL and libvips reading tiff, and libvips is so fast! But i encounter a problem. I can read pixel of each band through GDAL, but i don't know how to read each band through libvips, for example, here is the code i read tiff with GDAL.
```c++
	for (int bandind = 1; bandind <= bandcount; bandind++)
	{
		poBand = poDataset->GetRasterBand(bandind);	
		BYTE *pafScanline = new BYTE[nImgSizeX*nImgSizeY];
		poBand->RasterIO(GF_Read, 0, 0, nImgSizeX, nImgSizeY, pafScanline, nImgSizeX, nImgSizeY, GDT_Int16, 0, 0);
		for (int i = 0; i < nImgSizeX; i++)
		{
			for (int j = 0; j < nImgSizeY; j++)
			{
				cout << pafScanline[i*nImgSizeY + j] << endl;
			}
		}
         }
```
Here is the code i read tiff with libvips.
```c
            if (vips_region_prepare(region, &r))
                vips_error("vips_region_prepare error!", NULL);
     
            for (int y = 0; y < r.height; y++) {
                for (int x = 0; x < r.width; x++) {
                    printf("%3d ",*VIPS_REGION_ADDR(region, r.left + x, r.top + y));
                }  
                printf("\n");
            }
```
I can see libvips print pixels of the first band, how to get the other bands? Looking forward to your favourable reply! In addition, I want to donate some money to this project, but I can't use PayPal in China. Is there any other way to donate？

**Comment by jcupitt** [comment, marked as answer]
You can read bands like this:

```C
            if (vips_region_prepare(region, &r))
                vips_error("vips_region_prepare error!", NULL);
     
            for (int y = 0; y < r.height; y++) {
                for (int x = 0; x < r.width; x++) 
                    for (int b = 0; b < region->im->Bands; b++)
                        printf("%3d ",VIPS_REGION_ADDR(region, r.left + x, r.top + y)[b]);
             
                printf("\n");
            }
```

ie. libvips images are stored as RGBRGBRGBRGB (or however many bands there are) in memory.

You can make this a bit quicker:

```C
            if (vips_region_prepare(region, &r))
                vips_error("vips_region_prepare error!", NULL);
     
            for (int y = 0; y < r.height; y++) {
                VipsPel *p = VIPS_REGION_ADDR(region, r.left, r.top + y);

                for (int x = 0; x < r.width; x++) 
                    for (int b = 0; b < region->im->Bands; b++)
                        printf("%3d ", *p++);
             
                printf("\n");
            }
```

Since the `VIPS_REGION_ADDR()` macro includes a function call. You can use `VIPS_REGION_LSKIP()` (line skip) to find the row stride, which can speed it up a bit more.

https://libvips.github.io/libvips/API/current/VipsRegion.html#VIPS-REGION-LSKIP:CAPS

It's kind of you to offer to donate. Opencollective allows credit card payments, at least in the UK. Does that not work in China?

https://opencollective.com/libvips

**Reply by Zhaoyazhou1996** [reply]
Thank you for your reply! I have registered PayPal and i am waiting for audit, your helpful spirit is commendable! Have a good day!
